### DISCUSSION TITLE
Remove all external dependencies

### DISCUSSION THREAD

**Discussion by MagicLike** [JulianPrieber/littlelink-custom#222] [In progress] [answered]
Here is a list of all external dependencies of LLC (_subdomains cutted_):

- [x] fontawesome.com
- [x] getbootstrap.com
- [x] jsdelivr.net
- [x] cloudflare.com
- [x] googleapis.com
- [x] bootstrapcdn.com
- [x] jquery.com
- [x] bunny.net
- [x] duckduckgo.com

**Comment by JulianPrieber** [comment]
```
https://kit.fontawesome.com/c4a5e06183.js
https://use.fontawesome.com/releases/v6.1.2/css/all.css
https://pro.fontawesome.com/releases/v5.10.0/css/all.css
https://cdn.jsdelivr.net/npm/bootstrap-icons@1.8.1/font/bootstrap-icons.css
https://cdn.ckeditor.com/ckeditor5/35.1.0/classic/ckeditor.js
https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.slim.min.js
https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/js/bootstrap.bundle.min.js
https://cdnjs.cloudflare.com/ajax/libs/ace/1.4.1/ace.js
https://cdnjs.cloudflare.com/ajax/libs/vue/2.5.17/vue.js
https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js
https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js
//maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js
```

**Reply by JulianPrieber** [reply]
`https://code.jquery.com/jquery-1.12.4.min.js`

**Comment by JulianPrieber** [comment]
https://github.com/JulianPrieber/littlelink-custom/commit/56e17b9fd50cbc03ed2858bd2cca805a856be538

**Reply by JulianPrieber** [reply]
https://github.com/JulianPrieber/littlelink-custom/commit/ca2be62f8134afdd214d0592e818b431f9671011

**Comment by JulianPrieber** [comment, marked as answer]
Should we self-host [fonts.bunny.net](https://fonts.bunny.net/about)?

> What are Bunny Fonts?
> Bunny Fonts is an open-source, privacy-first web font platform designed to put privacy back into the internet.
> 
> With a zero-tracking and no-logging policy, Bunny Fonts helps you stay fully GDPR compliant and puts your user's personal data into their own hands. Additionally, you can enjoy lightning-fast load times thanks to bunny.net's global CDN network to help improve SEO and deliver a better user experience.

**Reply by My1** [reply]
biggest issue is obviously that it is a 3rd party and proving anything is difficult, I mean Google fonts recently put up a similar statement saying they dont log GFonts related stuff but I guess that likely wont matter in terms of people getting a cease and desist or whatever on that.

**Reply by JulianPrieber** [reply]
https://github.com/JulianPrieber/littlelink-custom/commit/5600ac090cdd4c7df48fac1a384ee4b9e5ce7e93

**Reply by JulianPrieber** [reply]
Most custom themes still use external resources, including Bunny Fonts.

**Comment by My1** [comment]
something on duckduckgo was also there, for website icons btw.

**Reply by JulianPrieber** [reply]
Unless someone comes up with an open source solution, we can't get around that.

**Reply by My1** [reply]
does it do anything more than favicons, because if it's only that LLC could possibly just download the favicon server-side and put up that

**Reply by JulianPrieber** [reply]
The thing is, most websites load favicons in a way that I have no idea how I would do that.

**Reply by My1** [reply]
oh I thought favicons generally get placed at /favicon.ico

**Reply by My1** [reply]
but even then couldnt the pull from duckduckgo happen server-side?

**Reply by JulianPrieber** [reply]
Maybe 20 years ago. Now, we can't assume that everyone does that.

**Reply by JulianPrieber** [reply]
The littlelink custom app itself even doesn't do that.

**Reply by JulianPrieber** [reply]
> but even then couldnt the pull from duckduckgo happen server-side?

On our system?

**Reply by My1** [reply]
> > but even then couldnt the pull from duckduckgo happen server-side?
> 
> On our system?

well basically the system of whoever's hosting.
and then instead of running an img tag to the 3rd party space, just img to that.
this could possibly even be expanded into a custom image upload someday.

**Reply by JulianPrieber** [reply]
What we could maybe do is self-host our own API using something like this: https://github.com/bbc/get-favicons. Or just proxy an existing API to anonymize the thing.

But even then we would rely on one of our servers. And this is a stretch if I can even get this thing going.

As it is now, I can't really see any way how I can integrate something like this into the app.

If I just write a system that checks for the /favicon.ico this would be a downgrade to our current system.
We already downgraded from the Google API, which worked amazingly but is hosted by Google...

I have people screaming to get back to the Google API, I can't justify making the system any worse than it is right now.
We need something that at least compares to the DuckDuckGo API and at best rivals the Google API so that everyone is happy.

**Reply by My1** [reply]
why should it need to rely on your servers? couldnt just LLC itself go out with curl or whatever, pull the image and store it locally for serving, that way if the user wants so even the google API wouldnt be a too big issue, as the only one putting out a request to google is the admin who is setting up the links and the image itself is just served from local

**Reply by JulianPrieber** [reply]
The only reason why I haven't implemented uploading custom icons is because I have to expand the database for that, which is an absolute nightmare to get working with the updater.

So storing anything locally isn't really an option until we update the updater to be able to expand existing database entries.

**Reply by My1** [reply]
I see but this could be a good 2 in one for the future, where you could upload from local and/or URL including some pre-fills for Google/DDG icon API, where one could have both fully custom icons as well as website icons from local

**Reply by JulianPrieber** [reply]
yeah

**Reply by JulianPrieber** [reply]
I found something online that had a good approach, I modified it for us to work, but it's not perfect. It works for some site, but not all.
The thing is using an interesting approach that isn't really efficient but good enough for us.
The thing is making a request with PHP and gets the site's source code. Then it searches for the bit that defines the favicon and display that source directly.

I made this quick API out of it, but it's just not good enough as it is right now to implement.
If we can get this working properly, I can include it locally.

https://components.littlelink-custom.com/icon/?https://littlelink-custom.com
https://components.littlelink-custom.com/icon/?https://www.apache.org

**Reply by JulianPrieber** [reply]
Something like this doesn't even work at all: https://components.littlelink-custom.com/icon/?https://demo.littlelink-custom.com/

**Reply by JulianPrieber** [reply]
@My1 
I think I got it to a state where it's better than the DuckDuckGo API:
https://components.littlelink-custom.com/ico/?https://demo.littlelink-custom.com
https://components.littlelink-custom.com/ico/?https://littlelink-custom.com
https://components.littlelink-custom.com/ico/?https://google.com
https://components.littlelink-custom.com/ico/?https://duckduckgo.com

I really gotta sort out some bugs before I can implement this.

**Reply by JulianPrieber** [reply]
The Google icon works locally in my test instance for me, but doesn't on the server. 🤔

**Reply by JulianPrieber** [reply]
I released a beta version that implements this. https://github.com/JulianPrieber/littlelink-custom/commit/16b4d652c92c35f91f06e660d7b0f496d2bf472f
